sda-commons-client-oidc/src/main/java/org/sdase/commons/client/oidc/OidcClient.java

```diff
@@ -0,0 +1,132 @@
+
+import static org.sdase.commons.client.jersey.proxy.ApiClientInvocationHandler.createProxy;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.concurrent.TimeUnit;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.core.Form;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import org.glassfish.jersey.client.proxy.WebResourceFactory;
+import org.sdase.commons.client.jersey.ClientFactory;
+import org.sdase.commons.client.oidc.model.*;
+import org.sdase.commons.client.oidc.rest.model.OpenIdDiscoveryResource;
+import org.sdase.commons.client.oidc.rest.model.TokenResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Service that can retrieve a new access token. */
+public class OidcClient {
+
+  private static final String CACHE_KEY = "oidcResultKey";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OidcClient.class);
+
+  private final Client client;
+
+  private final OidcConfiguration config;
+
+  private final OpenIdDiscoveryApi discoveryApi;
+
+  private Cache<String, TokenResource> cache;
+
+  public OidcClient(ClientFactory clientFactory, OidcConfiguration config) {
+    this.config = config;
+    if (config.isDisabled()) {
+      LOGGER.warn("OIDC was disabled.");
+      client = null;
+      discoveryApi = null;
+      return;
+    }
+
+    this.client =
+        clientFactory.externalClient(config.getHttpClient()).buildGenericClient("oidc-client");
+    this.discoveryApi =
+        createProxy(
+            OpenIdDiscoveryApi.class,
+            WebResourceFactory.newResource(
+                OpenIdDiscoveryApi.class, client.target(config.getIssuerUrl())));
+  }
+
+  /**
+   * Retrieves a new access token from the token endpoint using the given {@link OidcConfiguration}.
+   */
+  public OidcResult createAccessToken() {
+    if (config.isDisabled()) {
+      LOGGER.debug("OIDC client is disabled");
+      return new OidcResult().setState(OidcState.SKIPPED);
+    }
+
+    TokenResource tokenResource;
+
+    if (cache == null) {
+      tokenResource = getTokenResource();
+      if (tokenResource != null) {
+        cache =
+            Caffeine.newBuilder()
+                .expireAfterWrite(tokenResource.getAccessTokenExpiresInSeconds(), TimeUnit.SECONDS)
+                .maximumSize(100)
+                .build();
+        cache.put(CACHE_KEY, tokenResource);
+      }
+    } else {
+      tokenResource = cache.get(CACHE_KEY, k -> getTokenResource());
+    }
+    if (tokenResource != null) {
+      return new OidcResult().setState(OidcState.OK).setAccessToken(tokenResource.getAccessToken());
+    } else {
+
+      return new OidcResult().setState(OidcState.ERROR);
+    }
+  }
+
+  private TokenResource getTokenResource() {
```

Comment:
It would be even nicer to separate the actual business logic, i.e. token retrieval and the cache functionality.